src/core/insights/index.ts

```diff
@@ -0,0 +1,199 @@
+    posthog.disableDebugMode();
+  } catch (e) {
+    console.error("Failed to disable debug mode", e);
+  }
+};
+
+export const identify = ({
+  userId,
+  accountId,
+  organisationId,
+  email,
+  name,
+}: InsightsIdentity) => {
+  // In very rare cases we might have a user without an account, so we'll
+  // let null/undefined/blank strings through on that one
+  if (!userId) {
+    if (debugMode) {
+      console.warn("User ID not provided, skipping identify");
+    }
+    return;
+  }
+
+  try {
+    mixpanel.identify({ userId, accountId, organisationId, email, name });
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to identify user in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.identify({ userId, accountId, organisationId, email, name });
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to identify user in Posthog", e);
+    }
+  }
+};
+
+export const trackPageView = () => {
+  try {
+    mixpanel.trackPageView();
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track page view in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.trackPageView();
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track page view in Posthog", e);
+    }
+  }
+};
+
+export const track = (event: string, properties?: Record<string, unknown>) => {
+  try {
+    mixpanel.track(event, properties);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track event in Mixpanel", e);
+    }
+  }
+
+  try {
+    posthog.track(event, properties);
+  } catch (e) {
+    if (debugMode) {
+      console.error("Failed to track event in Posthog", e);
+    }
+  }
+};
+
+type InsightAttributes = {
+  event?: string;
+  [key: string]: string | undefined;
+};
+
+export const setupObserver = () => {
+  // Helper to get all data-insight-* attributes from an element
+  const getInsightAttributes = (element): InsightAttributes => {
+    // limit how many data attributes we'll process
+    const MAX_ATTRIBUTES = 10;
+    let count = 0;
+
+    const attributes: InsightAttributes = {};
+
+    for (const attr of element.attributes) {
+      if (count >= MAX_ATTRIBUTES) break;
+      if (attr.name.startsWith("data-insight-")) {
+        // Validate attribute name format
+        if (!/^data-insight-[a-zA-Z0-9-]+$/.test(attr.name)) continue;
+
+        // Sanitize attribute value
+        if (typeof attr.value !== "string" || attr.value.length > 100) continue;
+
+        // Convert data-insight-event-name to eventName
+        const key = attr.name
+          .replace("data-insight-", "")
+          .split("-")
+          .map((part, index) =>
+            index === 0 ? part : part.charAt(0).toUpperCase() + part.slice(1),
+          )
+          .join("");
+        attributes[key] = attr.value;
+        count++;
+      }
+    }
+    return attributes;
+  };
+
+  // Helper to find closest element with data-insight attributes
+  const findClosestElementWithInsights = (element) => {
```

Comment:
Nice way of dealing with ambiguous event propagation. I don't think it's realistically possible to click on the `<html>` element so the while loop should be fine